### PR TITLE
Add multiclass support

### PR DESCRIPTION
**Pull request by MichalBrzozowski91**

### CODE REVIEW

File: notebooks/requirements.txt
Comment (mwachnicki) [inline notebooks/requirements.txt:1]: Do we really need to require specific versions here?

File: tests/test_bert_truncated.py [hunk 1]
```diff
@@ -1,66 +1,60 @@
...
+import torch
+
 from belt_nlp.bert_truncated import BertClassifierTruncated
 
-MODEL_PARAMS = {
-    "batch_size": 1,
-    "learning_rate": 5e-5,
-    "epochs": 1,
-}
+MODEL_PARAMS = {"num_labels": 2, "batch_size": 1, "learning_rate": 5e-5, "epochs": 1, "device": "cuda"}
```
Comment (mwachnicki) [inline tests/test_bert_truncated.py:8]: Shouldn't we have `"device": "cpu"` by default as before?

**Comment by mwachnicki** [comment]
Summary of issues we have discussed today on the call:
* Divide classes into `Classifier` and `Regressor`
* Current `BertClassifier` class as a base for both
* Interface:
  * Classification: `predict` method returning classes and `predict_scores` method returning probabilities
  * Regression: `predict` method
* `_predict_logits` as a protected method
* Refactor names of classes
* Make sure that the cross entropy loss uses softmax and is the best choice in terms of numerical stability, etc.
  * Actually, the best option would be to have the loss function as a parameter with the default value
* Similarly, there can be a possibility to choose the optimizer and its parameters

File: notebooks/requirements.txt
Comment (MichalBrzozowski91) [reply notebooks/requirements.txt:1]: We do not - deleted

**Comment by MichalBrzozowski91** [comment]
Thanks for the review :)
I added the main changes concerning division classes into `Classifier` and `Regressor`
I did not implement the custom loss and optimizer:
- For me, it does violate the YAGNI principle - I am not sure this is needed now.
- If necessary, I would rather add it in another PR. This issue is separate from adding the multiclass support.

File: notebooks/requirements.txt
Comment (mwachnicki) [reply notebooks/requirements.txt:1]: We need a minimal version of datasets (e.g. `datasets>=2.17`). The `imdb` dataset didn't work for me with `2.12.0`.

File: notebooks/binary_classification/base.ipynb [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+   "execution_count": 2,
+   "id": "8f020349-c880-426d-a878-3089c79c9203",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "/home/mbrzozowski/projects/belt_add_multiclass/.venv/lib/python3.9/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
+      "  from .autonotebook import tqdm as notebook_tqdm\n"
```
Comment (mwachnicki) [inline notebooks/binary_classification/base.ipynb:30]: This warning can be dismissed by adding `ipywidgets` to the requirements file.

File: notebooks/multiclass/base.ipynb [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "id": "b8990378",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "dataset = load_dataset(\"guardian_authorship\",\"cross_topic_1\")"
```
Comment (mwachnicki) [inline notebooks/multiclass/base.ipynb:75]: A space is missing after the comma.

CI is not set up yet, but for now, you can check the errors with `nbqa flake8 .`

**Review by mwachnicki — CHANGES_REQUESTED** [review]
Thank you, I really like how it looks now! 🙂 

I agree that the loss and optimizer feature is not part of the multiclass support. I have just created the related issue - #43.

**Comment by mwachnicki** [comment]
I would also love to see the same basic tests for the Regressor classes.

File: belt_nlp/bert_with_pooling.py [hunk 4]
```diff
@@ -75,7 +78,7 @@ def __init__(
         self._params.update(additional_params)
 
         self.device = device
-        self.collate_fn = BertClassifierWithPooling.collate_fn_pooled_tokens
+        self.collate_fn = BertBaseWithPooling.collate_fn_pooled_tokens
```
[the remark this answers is not in the document]
Comment (MichalBrzozowski91) [reply belt_nlp/bert_with_pooling.py:81]: But this is is a static method, it seems natural to me to use with the class and not an instance.

File: belt_nlp/bert_regressor_truncated.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+            tokenizer=tokenizer,
+            neural_network=neural_network,
+            pretrained_model_name_or_path=pretrained_model_name_or_path,
+            device=device,
+            many_gpus=many_gpus,
+        )
+
+    def predict(self, x: list[str], batch_size: Optional[int] = None) -> list[int]:
+        """Returns regression scores."""
+        logits = super()._predict_logits(x, batch_size)
```
[the remark this answers is not in the document]
Comment (MichalBrzozowski91) [reply belt_nlp/bert_regressor_truncated.py:39]: I changed to returning Tensors.

File: belt_nlp/bert_with_pooling.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mwachnicki) [reply belt_nlp/bert_with_pooling.py:81]: You may be right in general, but this is a specific case, where the `collate_fn_pooled_tokens` method is called in an instance method of the same class.

If you create a new class that derives from the `BertBaseWithPooling` class and overwrite the `collate_fn_pooled_tokens` method, you would probably expect the overwritten method to be used in the `__init__` method and that's why I think it would be safer to call it as `self.collate_fn_pooled_tokens`.

However, in this very specific case, this scenario may be unlikely, so I just want to present you my point of view, but I leave the decision to you.

File: tests/test_bert_regressor_truncated.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+def test_regression_order():
+    """Check if the order of predictions is preserved."""
+    params = MODEL_PARAMS
+    model = BertRegressorTruncated(**params)
+    x_train = ["good", "super", "ok", "great"]
+    y_train = [1] * len(x_train)
+
+    x_test = ["nice"] * 99 + ["bad"] * 1
+
+    model.fit(x_train, y_train)
```
Comment (mwachnicki) [inline tests/test_bert_regressor_truncated.py:34]: Right now, `y_train` expects `list[bool]`, so the type hint should be updated.

File: tests/test_bert_regressor_truncated.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+    x_test = ["nice", "cool"]
+
+    model.fit(x_train, y_train)
+    scores = model.predict(x_test)
+
+    assert scores.shape == torch.Size([2, 1])
+
+
+def test_regression_order():
```
Comment (mwachnicki) [inline tests/test_bert_regressor_truncated.py:25]: I think that for both classifiers and regressors `prediction_order` is the appropriate name.

`predictions` is used in the docstring anyway ;)

**Review by mwachnicki — CHANGES_REQUESTED** [review]

File: belt_nlp/bert_with_pooling.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (MichalBrzozowski91) [reply belt_nlp/bert_with_pooling.py:81]: Thanks for clarification. Yes, you are right, I will change it.

**Review by mwachnicki — APPROVED** [review]
Good work, well done!

Remember about updating the date in the changelog and pushing the new version to PyPI :wink:
